File: CSharp/SpecStudioExamplesTestsCSharp/common/DiscountingTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class DiscountingTyped
    {
        public Dollar totalPrice;
        public Percentage discount;
        public string notes;

        public DiscountingTyped(Dollar totalPrice, Percentage discount, string notes)
        {
            this.totalPrice = totalPrice;
            this.discount = discount;
            this.notes = notes;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteString("totalPrice", Json.ToText(this.totalPrice));
            w.WriteString("discount", Json.ToText(this.discount));
            w.WriteString("notes", this.notes);
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static DiscountingTyped FromJsonElement(JsonElement m)
        {
            return new DiscountingTyped(
                new Dollar(Json.AsString(Json.Require(m, "totalPrice"), "totalPrice")),
                new Percentage(Json.AsString(Json.Require(m, "discount"), "discount")),
                Json.AsString(Json.Require(m, "notes"), "notes")
            );
        }

        public static DiscountingTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<DiscountingTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<DiscountingTyped> FromJSONList(string json)
        {
            var result = new List<DiscountingTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "DiscountingTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }

        public override string ToString()
        {
            return $"Total Price={totalPrice}, Discount={discount}, Notes={notes}";
        }

        public override bool Equals(object? obj)
        {
            if (obj is not DiscountingTyped other) return false;
            return object.Equals(this.totalPrice, other.totalPrice)
                && object.Equals(this.discount, other.discount)
                && object.Equals(this.notes, other.notes);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.totalPrice);
            h.Add(this.discount);
            h.Add(this.notes);
            return h.ToHashCode();
        }
    }
}

File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/DiscountingTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;

public class DiscountingTyped {
    public Dollar totalPrice;
    public Percentage discount;
    public String notes;

    public DiscountingTyped(Dollar totalPrice, Percentage discount, String notes) {
        this.totalPrice = totalPrice;
        this.discount = discount;
        this.notes = notes;
    }

    public DiscountingTyped(DiscountingString s) {
        this.totalPrice = new Dollar(s.totalPrice);
        this.discount = new Percentage(s.discount);
        this.notes = s.notes;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof DiscountingTyped)) return false;
        DiscountingTyped that = (DiscountingTyped) o;
        return Objects.equals(totalPrice, that.totalPrice)
            && Objects.equals(discount, that.discount)
            && Objects.equals(notes, that.notes);
    }

    @Override
    public int hashCode() {
        return Objects.hash(totalPrice, discount, notes);
    }

    public DiscountingString toDiscountingString() {
        return new DiscountingString(String.valueOf(totalPrice), String.valueOf(discount), String.valueOf(notes));
    }

    public static List<DiscountingTyped> fromStringList(List<DiscountingString> list) {
        List<DiscountingTyped> result = new ArrayList<>();
        for (DiscountingString s : list) result.add(new DiscountingTyped(s));
        return result;
    }

    public static List<DiscountingString> toStringList(List<DiscountingTyped> list) {
        List<DiscountingString> result = new ArrayList<>();
        for (DiscountingTyped t : list) result.add(t.toDiscountingString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("totalPrice", totalPrice == null ? null : totalPrice.value);
        m.put("discount", discount == null ? null : discount.value);
        m.put("notes", notes);
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static DiscountingTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new DiscountingTyped(
            new Dollar(Json.asString(Json.require(m, "totalPrice"), "totalPrice")),
            new Percentage(Json.asString(Json.require(m, "discount"), "discount")),
            Json.asString(Json.require(m, "notes"), "notes"));
    }

    public static DiscountingTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<DiscountingTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (DiscountingTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<DiscountingTyped> fromJSONList(String json) {
        List<DiscountingTyped> result = new ArrayList<DiscountingTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "DiscountingTyped")));
        return result;
    }
}
